src/AdoNet/Orleans.Persistence.AdoNet/Storage/Provider/Orleans3CompatibleStringKeyHasher.cs

```diff
@@ -37,17 +38,63 @@ public int Hash(byte[] data)
             // PickHasher parameters are the same for both calls so we need to analyze data content to distinguish these cases.
             // It doesn't word if string key is equal to grain type name, but we consider this edge case to be negligibly rare.
 
-            // reducing allocations if data is not a grain type
-            if (data.Length >= _grainType.Length && Encoding.UTF8.GetByteCount(_grainType) == data.Length)
+            if (IsGrainTypeName(data))
+                return _innerHasher.Hash(data);
+
+            var extendedLength = data.Length + 8;
+            if (extendedLength <= 256)
+            {
+                Span<byte> extended = stackalloc byte[extendedLength];
+                data.AsSpan().CopyTo(extended);
+                return _innerHasher.Hash(extended);
+            }
+
+            var buffer = ArrayPool<byte>.Shared.Rent(extendedLength);
+            try
+            {
+                data.AsSpan().CopyTo(buffer);
+                Array.Clear(buffer, data.Length, 8);
+                return _innerHasher.Hash(buffer.AsSpan(0, extendedLength));
+            }
+            finally
+            {
+                ArrayPool<byte>.Shared.Return(buffer);
+            }
+        }
+
+        private bool IsGrainTypeName(byte[] data)
+        {
+            // at least 1 byte per char
+            if (data.Length < _grainType.Length)
+                return false;
+
+            var grainTypeByteCount = Encoding.UTF8.GetByteCount(_grainType);
+            if (grainTypeByteCount != data.Length)
+                return false;
+
+            if (grainTypeByteCount <= 256)
             {
-                var grainTypeBytes = Encoding.UTF8.GetBytes(_grainType);
-                if (grainTypeBytes.AsSpan().SequenceEqual(data))
-                    return _innerHasher.Hash(data);
+                Span<byte> grainTypeBytes = stackalloc byte[grainTypeByteCount];
+                if (!Encoding.UTF8.TryGetBytes(_grainType, grainTypeBytes, out _))
+                    throw new InvalidOperationException();
+
+                return grainTypeBytes.SequenceEqual(data);
             }
 
-            var extendedData = new byte[data.Length + 8];
-            data.CopyTo(extendedData, 0);
-            return _innerHasher.Hash(extendedData);
+            var buffer = ArrayPool<byte>.Shared.Rent(grainTypeByteCount);
+            try
+            {
+                var grainTypeBytes = buffer.AsSpan(0, grainTypeByteCount);
+
+                if (!Encoding.UTF8.TryGetBytes(_grainType, grainTypeBytes, out _))
+                    throw new InvalidOperationException();
+
+                return grainTypeBytes.SequenceEqual(data);
+            }
+            finally
```

Comment:
`finally` isn't needed. In case of failure the rented array will just be dropped (on the GCed). The array-pool is tolerant to this.
W/o `finally` there's also room for the JIT to do more optimizations.

The pattern w/ `try-finally` for the array pool got used some time ago in .NET, but it isn't anymore.
(Same above)